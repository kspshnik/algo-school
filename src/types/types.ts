import {
  ASC, BUBBLE_SORT, DSC, INSERTION_SORT, NONE,
} from '../constants/type-literals';
import { TSortStepResult, TStringStepResult } from './algo-struct.types';
import Direction from './direction';

export type TSortingChoiceType = typeof BUBBLE_SORT | typeof INSERTION_SORT | typeof NONE;

export type TSortingType = typeof BUBBLE_SORT | typeof INSERTION_SORT;

export interface SortingGeneratorInterface {
  (arr : Array<number>, direction : Direction) : Generator<TSortStepResult, void, never>,
}

export type TSortGeneratorsObject = {
  [key in TSortingType] : SortingGeneratorInterface;
};

export type TSortingDirection = typeof ASC | typeof DSC | typeof NONE;

export type TAlgoData = string | Array<number>;

export interface AlgorithmsGenerator {
  (data : TAlgoData) : AlgorithmsIterator;
}

// eslint-disable-next-line max-len
export type AlgorithmsIterator = Generator<TStringStepResult | Array<number> | TSortStepResult, void, never>;

export interface AlgorithmIterationHookInterface {
  (generator : AlgorithmsGenerator, data : TAlgoData) : React.MutableRefObject<AlgorithmsIterator>,
}
