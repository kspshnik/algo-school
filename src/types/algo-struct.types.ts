import QueueNode from '../services/data-structures/queue-node';

export interface StringReverseGeneratorInterface {
  (string : string, head : number, tail : number) : string,
}

export type TStringStepResult = {
  arr : Array<string>,
  changing : Array<number>,
  ready : Array<number>,
};

export type StringAlgorithmIteratorInterface = Generator<TStringStepResult, void, never>;
export type FibonacchiAlgorithmIteratorInterface = Generator<Array<number>, void, never>;
export type SortingAlgorithmIteratorInterface = Generator<TSortStepResult, void, never>;

export type TSortStepResult = {
  active : Array<number>,
  ready : Array<number>,
  array : Array<number>,
};

export interface AbstractNodeInterface<T> {
  value : T,
}

export interface StackNodeInterface<T> extends AbstractNodeInterface<T> {
  on : StackNodeInterface<unknown> | null,

  remove() : void;
}

export interface StackInterface {
  top : StackNodeInterface<unknown> | null,
  isEmpty : boolean,

  push(node : StackNodeInterface<unknown>) : void,

  pop() : StackNodeInterface<unknown> | null,

  peek() : StackNodeInterface<unknown> | null,

  purge() : void,
}

export interface QueueNodeInterface<T> extends AbstractNodeInterface<T> {
  prev : QueueNode<unknown> | null;

  remove() : void;
}
