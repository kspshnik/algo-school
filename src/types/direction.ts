enum Direction {
  Ascending = 'ascending',
  Descending = 'descending',
}

export default Direction;
