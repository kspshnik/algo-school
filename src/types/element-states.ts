// eslint-disable-next-line import/prefer-default-export
export enum ElementStates {
  Default = 'default',
  Changing = 'changing',
  Modified = 'modified',
}
