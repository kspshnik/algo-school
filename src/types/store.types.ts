import { TSortingChoiceType } from './types';
import { Direction } from './direction';

export type TAlgoViewItem = {
  id : string,
  isChanging : boolean,
  isDone : boolean,
  value : string | number,
};

export type TStructViewItem = [
  TAlgoViewItem | string | null,
  TAlgoViewItem,
  TAlgoViewItem | string | null,
];
export type TStructView = Array<TStructViewItem>;
export type TAlgoView = Array<TAlgoViewItem>;

export type TAlgoSliceState = {
  viewData : TAlgoView,
  isFinished : boolean,
  isActive : boolean,
};

export type TStructSliceState = {
  viewData : TStructView,
  isFinished : boolean,
  isActive : boolean,
  start : number,
  end: number,
};
export type TFormsSliceState = {
  string : string,
  fibonacchiLimit : number,
  randomArray : Array<number>,
  sortingType : TSortingChoiceType,
  direction : Direction | null,
  item : string,
  index : number,
};
