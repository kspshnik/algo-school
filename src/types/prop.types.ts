import React from 'react';
import { ElementStates } from './element-states';
import { Direction } from './direction';

export interface PageLayoutProps {
  title : string;
  extraClass? : string;
}

export interface MainPageProps {
  extraClass? : string;
}

export interface CircleProps {
  state? : ElementStates,
  letter? : string,
  head? : string | React.ReactElement | null,
  index? : number,
  tail? : string | React.ReactElement | null,
  tailType? : 'string' | 'element',
  extraClass? : string,
  isSmall? : boolean,
  onAnimationEnd : React.AnimationEventHandler,
}

export interface ColumnProps {
  index : number;
  state? : ElementStates;
  extraClass? : string;
}

export interface ButtonProps extends React.HTMLProps<HTMLButtonElement> {
  text? : string;
  type? : 'button' | 'submit' | 'reset';
  sorting? : Direction;
  linkedList? : 'small' | 'big';
  isLoader? : boolean;
  extraClass? : string;
}

export interface ArrowIconProps {
  fill? : string;
}

export interface ReturnButtonProps extends React.HTMLProps<HTMLButtonElement> {
  type? : 'button' | 'submit' | 'reset';
  extraClass? : string;
}

export interface InputProps extends React.HTMLProps<HTMLInputElement> {
  placeholder? : string;
  extraClass? : string;
  isLimitText? : boolean;
}

export interface RadioInputProps extends React.HTMLProps<HTMLInputElement> {
  label : string;
  extraClass? : string;
}

export interface LayoutProps {
  children : React.ReactNode | Array<React.ReactNode>,
}
