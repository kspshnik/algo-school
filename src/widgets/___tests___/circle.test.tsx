/* eslint-disable react/react-in-jsx-scope */
import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { Circle } from '../index';
import { ElementStates } from '../../types/element-states';

describe('Circle widget renders in a right way', () => {
  it('should render', () => {
    render(<Circle letter='ква' state={ElementStates.Changing} />);
    //  screen.debug();
    const circle = screen.getByTestId('widget-circle-ква');
    // console.dir(circle);
    expect(circle).toBeInTheDocument();
  });
  it('div should have class "changing', () => {
    render(<Circle letter='ура' state={ElementStates.Changing} />);
    const circle = screen.getByTestId('widget-circle-ура');
    expect(circle).toHaveClass('changing');
  });
  it('div should have class "default"', () => {
    render(<Circle letter='сыр' state={ElementStates.Default} />);
    const circle = screen.getByTestId('widget-circle-сыр');
    expect(circle).toHaveClass('default');
  });
  it('div should have class "modified"', () => {
    render(<Circle letter='соль' state={ElementStates.Modified} />);
    const circle = screen.getByTestId('widget-circle-соль');
    expect(circle).toHaveClass('modified');
  });
  it('should render tail with text', () => {
    render(<Circle head='шапка' tail='хвост' letter='фыр' />);
    const circle = screen.getByTestId('widget-circle-фыр');
    const tail = screen.getByText('хвост');
    expect(circle).toBeInTheDocument();
    expect(tail).toBeInTheDocument();
  });
  it('should render head  with text', () => {
    render(<Circle head='шапка' tail='хвост' letter='фыр' />);
    const circle = screen.getByTestId('widget-circle-фыр');
    const head = screen.getByText('шапка');
    expect(circle).toBeInTheDocument();
    expect(head).toBeInTheDocument();
  });
  it('should render small circle in the head', () => {
    render(<Circle head={<Circle letter='куку' isSmall />} letter='быр' />);
    const circle = screen.getByTestId('widget-circle-быр');
    const head = screen.getByTestId('widget-circle-куку');
    expect(circle).toBeInTheDocument();
    expect(head).toBeInTheDocument();
  });
  it('should render small circle in the tail', () => {
    render(<Circle tail={<Circle letter='куку' isSmall />} letter='быр' />);
    const circle = screen.getByTestId('widget-circle-быр');
    const tail = screen.getByTestId('widget-circle-куку');
    expect(circle).toBeInTheDocument();
    expect(tail).toBeInTheDocument();
  });
});

describe('Circle widget should match snapshot in all renders', () => {
  it('should match snapshot', () => {
    const circleSnapshot = renderer.create(<Circle letter='ква' state={ElementStates.Default} />).toJSON();
    expect(circleSnapshot).toMatchSnapshot();
  });
  it('changing state should match snapshot', () => {
    const circleSnapshot = renderer.create(<Circle letter='ква' state={ElementStates.Changing} />).toJSON();
    expect(circleSnapshot).toMatchSnapshot();
  });
  it('done state should match snapshot', () => {
    const circleSnapshot = renderer.create(<Circle letter='ква' state={ElementStates.Modified} />).toJSON();
    expect(circleSnapshot).toMatchSnapshot();
  });
  it('circle enriched by head & tail both with text should match snapshot', () => {
    const circleSnapshot = renderer.create(<Circle head='шапка' tail='хвост' letter='фыр' />).toJSON();
    expect(circleSnapshot).toMatchSnapshot();
  });
  it('circle w/small Circle in head should match snapshot', () => {
    const circleSnapshot = renderer.create(<Circle head={<Circle letter='куку' isSmall />} letter='быр' />).toJSON();
    expect(circleSnapshot).toMatchSnapshot();
  });
  it('circle w/small Circle in tail should match snapshot', () => {
    const circleSnapshot = renderer.create(<Circle tail={<Circle letter='куку' isSmall />} letter='быр' />).toJSON();
    expect(circleSnapshot).toMatchSnapshot();
  });
});
