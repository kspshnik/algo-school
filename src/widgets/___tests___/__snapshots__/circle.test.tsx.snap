// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Circle widget should match snapshot in all renders changing state should match snapshot 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle   changing"
    data-testid="widget-circle-ква"
  >
    <p
      className="text text_type_circle text_color_input letter"
    >
      ква
    </p>
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;

exports[`Circle widget should match snapshot in all renders circle enriched by head & tail both with text should match snapshot 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head string"
  >
    шапка
  </div>
  <div
    className="circle   default"
    data-testid="widget-circle-фыр"
  >
    <p
      className="text text_type_circle text_color_input letter"
    >
      фыр
    </p>
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 string"
  >
    хвост
  </div>
</div>
`;

exports[`Circle widget should match snapshot in all renders circle w/small Circle in head should match snapshot 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  >
    <div
      className="content "
    >
      <div
        className="text text_type_input text_color_input mb-4 absolute head element"
      />
      <div
        className="circle  small default"
        data-testid="widget-circle-куку"
      >
        <p
          className="text text_type_circle text_color_input letter"
        >
          куку
        </p>
      </div>
      <p
        className="text text_type_input text_color_input mt-4 absolute index"
      />
      <div
        className="text text_type_input text_color_input mt-4 absolute tail30 element"
      />
    </div>
  </div>
  <div
    className="circle   default"
    data-testid="widget-circle-быр"
  >
    <p
      className="text text_type_circle text_color_input letter"
    >
      быр
    </p>
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;

exports[`Circle widget should match snapshot in all renders circle w/small Circle in tail should match snapshot 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle   default"
    data-testid="widget-circle-быр"
  >
    <p
      className="text text_type_circle text_color_input letter"
    >
      быр
    </p>
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  >
    <div
      className="content "
    >
      <div
        className="text text_type_input text_color_input mb-4 absolute head element"
      />
      <div
        className="circle  small default"
        data-testid="widget-circle-куку"
      >
        <p
          className="text text_type_circle text_color_input letter"
        >
          куку
        </p>
      </div>
      <p
        className="text text_type_input text_color_input mt-4 absolute index"
      />
      <div
        className="text text_type_input text_color_input mt-4 absolute tail30 element"
      />
    </div>
  </div>
</div>
`;

exports[`Circle widget should match snapshot in all renders done state should match snapshot 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle   modified"
    data-testid="widget-circle-ква"
  >
    <p
      className="text text_type_circle text_color_input letter"
    >
      ква
    </p>
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;

exports[`Circle widget should match snapshot in all renders should match snapshot 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle   default"
    data-testid="widget-circle-ква"
  >
    <p
      className="text text_type_circle text_color_input letter"
    >
      ква
    </p>
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;
